core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -77,21 +83,25 @@
   @Nullable
   private final Throttle throttle;
 
+  private Set<Substream> activeHedges; // not null and unchanged once isHedging = true
```

Comment:
I don't see any comprehensive thread-safety story. How is this class safe?

The synchronization involved between these new variables is really hard to follow. There are _4_ (actually 5 if you consider the timer) new independently synchronized variables.

Maybe it's also that it isn't clear the behavior of these variables. Like `activeHedges` I see things like checking `size() == 1` and iteration and it isn't clear to me how those are safe, logically. I understand they would be synchronized by the `Set`, but I don't understand how you can safely combine together logical conclusions if other threads could be performing concurrent modification of the set.

There is one new variable that _isn't_ synchronized: `latestSubstream`. However `hasPotentialHedging()` reads `latestSubstream` from the closed() callback, so it seems that it _should_ be synchronized because a new stream can start at any time.

Reply:
`activeHedges` and `rescheduler` can be regarded as final fields. (The `RetriableStream.cancel()` thread may read stale values of them but it's not big of a deal because things will still be cleaned up soon)

`activeHedges.size() == 1` is only checked after `hasPotentialHedging() == false`, which means absolutely no new hedge will start. (The only one left active hedge can be closed in a race, but the RPC can still get committed whichever thread wins)

The value `latestSubstream` may be stale when calling `hasPotentialHedging()`, and thus `hasPotentialHedging()` returns `true` (which could have been `false` if strongly synchronized); but it doesn't matter (you can regard that `hasPotentialHedging()` were called before the real `latestSubstream` was born), at least one other hedge will observe `hasPotentialHedging()` return `false` when they are `closed()`.

Reply:
> activeHedges and rescheduler can be regarded as final fields

I wasn't speaking as much to their reference, but to their contents. They both maintain mutating state.

> activeHedges.size() == 1 is only checked after hasPotentialHedging() == false, which means absolutely no new hedge will start.

Let's say there's three hedges and they all fail at the same time. `state.winningSubstream == null` could be `true` for all three hedges. It's a bit hard to tell, but it seems `hasPotentialHedging()` could end up being true for two of those hedges.

I see another race with `hedgingFrozen`. It seems a hedge could be starting _right now_ within `HedgingRunnable`, it could see `hedgingFrozen == false`, then hedging is frozen because of `RpcProgress.DROPPED`, and then `HedgingRunnable` calls `createSubstream()` which will do `activeHedges.add()`.

Logic-ing out this sort of thing is _really hard_ and _error prone_. Doing that for one or two variables is many times necessary, but here we are combining `state`, a _different_ `state` within `hasPotentialHedging()`, unsynchronized `latestSubstream`, `hedgingFrozen`, and `activeHedges`.

> The value latestSubstream may be stale when calling hasPotentialHedging()

This is something _really_ important to call out in the code. But also, I know Carl won't be happy about that because the race detector will complain about it.

Reply:
* `activeHedges`: trash. Let's not bother with early commitment. Because of transparent retries, most RPCs won't be able to early commit, so it's no use optimizing.
* `hedgingFrozen`: Merge into `state`. Could actually be handled by making previousAttempts really large. But I don't know if that is most clear in the code.
* `latestSubstream.previousAttempts`: Merge into `state`
* `latestSubstream.scheduledRetry`: since this is a pre-existing race and not related to hedging, let's handle this separately. So undo from this PR (TODO)
* `noMoreTransparentRetry`: check with spec authors whether we can do per-attempt transparent retry. If so, move this to `Substream`.

TODO (after first pass of simplification): interactions between `rescheduler` and `hedgingFrozen`/`previousAttempts`.